### PR TITLE
[WIP]: Scraper to generalise the downloads from http/ftp directories

### PR DESCRIPTION
**Pull request by dpshelio**
This is a Work in Progress to create a single scraper which could be used to download data from any site, avoiding therefore to create custom web scrappers for each data service.

For example, a way to use it within Callisto would be:
```python
pattern = "http://soleil.i4ds.ch/solarradio/data/2002-20yy_Callisto/%Y/%m/%d/{observatory}_%Y%m%d_%H%M%S_{number}.fit.gz"
scrap = Scraper(pattern, observatory='BIR', number='01')
urls = scrap.filelist(TimeRange(start, end))
```
or for goes:
```python
pattern = "http://umbra.nascom.nasa.gov/goes/fits/%Y/go{sat_num:02d}%Y%m%d.fits"
scrap = Scraper(pattern, sat_num = sat_num[0])
urls = scrap.filelist(TimeRange(start, end))
```
So far (and not in this PR) I've just tried to implement it in Callisto, however I haven't figured out yet how to make to take more than one `number`, as I cannot write in the current status a list of possible inputs (eg. `number = 01 || 03`)

Potentially we could then have a list of patterns and the general downloader in plan could choose the one to use.

Please, comments, criticisms, help is appreciated :)

### CODE REVIEW

**Comment by dpshelio** [comment]
I've updated this PR with the things suggested, documentation and updated the tests.  
This tool is extremely useful for two of the projects being now executed, now for VALS (@tortoiseJess) and sep for GSOC (@ankitkmr) as with this scraper we don't only generate one file per directory but it find all files available.

Please, review and comment :)

File: sunpy/util/scraper.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+import datetime
+import re
+
+import urllib2
+from bs4 import BeautifulSoup
+
+__all__ = ['Scraper']
+
+class Scraper:
```
Comment (Cadair) [inline sunpy/util/scraper.py]: should be `Scraper(object)`

File: sunpy/util/scraper.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        Returns
+        -------
+
+        directories : list of strings
+            List of all the possible directories valid for the time
+            range given. Notice that these directories may not exist
+            in the archive.
+        """
+        #find directory structure - without file names
+        directorypattern = self.pattern[:-self.pattern[::-1].find('/')]
```
Comment (Cadair) [inline sunpy/util/scraper.py]: you could replace this with `os.path.dirname`?

File: sunpy/util/scraper.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        #TODO what if there's not slashes?
+        rangedelta = timerange.dt # TODO check it's a timerange
```
Comment (Cadair) [inline sunpy/util/scraper.py]: you don't need the type check, any object that has a dt attribute will suffice.

File: sunpy/util/scraper.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        for k,v in time_conversions.iteritems():
+            pattern = pattern.replace(k, v)
+        matches = re.match(pattern, url)
+        if matches:
+            return matches.end() == matches.endpos == len(self.now)
+        return False
+
+    def _extractDateURL(self, url):
+        """Extracts the date from a particular url following the pattern"""
+        url_to_list = lambda txt: re.sub(r'\.|_', '/', txt).split('/')
```
Comment (Cadair) [inline sunpy/util/scraper.py]: can you write a layman comment explaining wtf that regex is doing?

File: sunpy/util/tests/test_scraper.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+from sunpy.time import TimeRange
+from sunpy.util.scraper import Scraper
+
+PATTERN_EXAMPLES = [
+    ('%b%y', datetime.timedelta(days=31)),
+    ('%m%y', datetime.timedelta(days=31)),
+    ('%H%d', datetime.timedelta(hours=1)),
+    ]
+
+class TestScraper:
```
Comment (Cadair) [inline sunpy/util/tests/test_scraper.py]: do you need the class? You don't seem to be sharing any variables etc? Just functions is probably better?

File: sunpy/util/tests/test_scraper.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        timerange = TimeRange(startdate, enddate)
+        assert len(s.filelist(timerange)) == 0
+
+
+    
+
+               
+
+
+        #TODO: data gaps
```
Comment (Cadair) [inline sunpy/util/tests/test_scraper.py]: needs tidying.

**Comment by Cadair** [comment]
This looks great to me. Just a few small things.

File: sunpy/util/scraper.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        #Find in directory and filename
+        for pattern_elem, url_elem in zip(pattern_list, url_list):
+            time_formats = [x for x in time_order if x in pattern_elem]
+            if len(time_formats) > 0:
+                final_date.append(url_elem)
+                final_pattern.append(pattern_elem)
+                for time_bit in time_formats:
+                    time_order.remove(time_bit)
+        return datetime.datetime.strptime(' '.join(final_date),
+                                          ' '.join(final_pattern))
```
Comment (ankitkmr) [inline sunpy/util/scraper.py]: I have a url which goes like : http://stereo-ssc.nascom.nasa.gov/data/ins_data/plastic/level2/Protons/ASCII/10min/A/2008/STA_L2_PLA_1DMax_10min_20080101_001_V09.txt and hence cause group name redefinition error in return above. 

We get something like this:

final_pattern = ['%Y', '%Y%m%d', '%j']
final_date = ['2008', '20080101', '001']

So please change last return line to :

try:
    print datetime.datetime.strptime(' '.join(final_date), ' '.join(final_pattern))

except:
    anomaly = [x for x in final_pattern if len(x) > 2]
    for x in anomaly:
        splits = [x[i:i+2] for i in range(0, len(x), 2)]
        index_remove = [final_pattern.index(i) for i in splits if i in final_pattern]
        for i in splits:
            if i in final_pattern: final_pattern.remove(i)
                
        for i in index_remove: final_date.remove(final_date[i])
            

    print datetime.datetime.strptime(' '.join(final_date), ' '.join(final_pattern))

File: sunpy/util/scraper.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        >>> print(solmon.filelist(timerange))
+        ['http://solarmonitor.org/data/2015/01/01/fits/swap/swap_00174_fd_20150101_025423.fts.gz']
+        """
+        directories = self.range(timerange)
+        filesurls = []
+        for directory in directories:
+            try:
+                opn = urllib2.urlopen(directory)
+                try:
+                    soup = BeautifulSoup(opn)
```
Comment (Cadair) [inline sunpy/util/scraper.py]: does this support context manger?

i.e.
```python
with BeautifulSoup(opn) as soup:
    for link...
```

Comment (dpshelio) [inline sunpy/util/scraper.py]: humm... that doesn't seem to work.  Or at least not by just changing that :(

I get an 
```python
AttributeError: __exit__
```

I've also tried to change it in the line above:
```python
with urllib2.urlopen(directory) as opn:
    ...
```
but also I get an 
```python
AttributeError: addinfourl instance has no attribute '__exit__'
```

Comment (Cadair) [inline sunpy/util/scraper.py]: fair enough. So that's a no then.

**Comment by ehsteve** [comment]
:+1:

**Comment by wafels** [comment]
:+1:

**Pull request merged**